File: temp_pose_engine/pose_engine/lib/services/mediapipe_pose_detector.dart
import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:flutter/foundation.dart';
import 'pose_detector_service.dart';
import '../models/pose_landmark.dart';
import '../models/pose_result.dart';

/// MediaPipe pose detector that connects to Python backend via WebSocket
/// 
/// This detector communicates with the Python MediaPipe server to get
/// real-time pose detection data from the camera.
class MediaPipePoseDetector implements PoseDetectorService {
  WebSocketChannel? _channel;
  bool _isReady = false;
  int _frameCount = 0;
  
  final String _host;
  final int _port;
  
  // Stream controller for pose results
  final StreamController<PoseResult> _poseStreamController = 
      StreamController<PoseResult>.broadcast();
  
  // Latest pose result for synchronous access
  PoseResult? _latestPose;
  
  // Connection status callback
  Function(bool connected, String message)? onConnectionStatus;
  
  MediaPipePoseDetector({
    String host = 'localhost',
    int port = 8765,
  }) : _host = host, _port = port;

  @override
  String get name => 'MediaPipePoseDetector (Python Backend)';

  @override
  bool get isReady => _isReady;
  
  /// Stream of pose detection results
  Stream<PoseResult> get poseStream => _poseStreamController.stream;

  @override
  Future<void> initialize() async {
    try {
      final uri = Uri.parse('ws://$_host:$_port');
      debugPrint('🔌 Connecting to MediaPipe server at $uri');
      
      _channel = WebSocketChannel.connect(uri);
      
      // Wait for connection
      await _channel!.ready;
      
      debugPrint('✅ Connected to MediaPipe server');
      
      // Listen for messages
      _channel!.stream.listen(
        _handleMessage,
        onError: (error) {
          debugPrint('❌ WebSocket error: $error');
          _isReady = false;
          onConnectionStatus?.call(false, 'Connection error: $error');
        },
        onDone: () {
          debugPrint('🔌 WebSocket connection closed');
          _isReady = false;
          onConnectionStatus?.call(false, 'Connection closed');
        },
      );
      
      // Send start command to begin camera capture on Python side
      _sendCommand({'command': 'start', 'camera_index': 0});
      
      _isReady = true;
      onConnectionStatus?.call(true, 'Connected to MediaPipe server');
      
    } catch (e) {
      debugPrint('❌ Failed to connect to MediaPipe server: $e');
      _isReady = false;
      onConnectionStatus?.call(false, 'Failed to connect: $e');
      rethrow;
    }
  }
  
  void _sendCommand(Map<String, dynamic> command) {
    if (_channel != null) {
      _channel!.sink.add(jsonEncode(command));
    }
  }
  
  // Stream controller for video frames (base64 encoded)
  final StreamController<String> _frameStreamController = 
      StreamController<String>.broadcast();
      
  /// Stream of video frames (base64 encoded JPEG)
  Stream<String> get frameStream => _frameStreamController.stream;

  void _handleMessage(dynamic message) {
    try {
      final data = jsonDecode(message as String);
      final type = data['type'] as String?;
      
      switch (type) {
        case 'frame':
          // Handle both pose and image
          if (data['pose'] != null) {
            _handlePoseData(data['pose']);
          }
          if (data['image'] != null) {
            _frameStreamController.add(data['image'] as String);
          }
          break;
        case 'pose':
          _handlePoseData(data['data']);
          break;
        case 'status':
          final success = data['success'] as bool;
          final msg = data['message'] as String;
          debugPrint('📊 Status: $msg');
          onConnectionStatus?.call(success, msg);
          break;
        case 'pong':
          debugPrint('🏓 Pong received');
          break;
      }
    } catch (e) {
      debugPrint('Error parsing message: $e');
    }
  }
  
  void _handlePoseData(Map<String, dynamic> poseData) {
    final detected = poseData['detected'] as bool? ?? false;
    final landmarks = <PoseLandmarkType, PoseLandmark>{};
    
    if (detected && poseData['landmarks'] != null) {
      final landmarkList = poseData['landmarks'] as List;
      
      for (final lm in landmarkList) {
        final index = lm['index'] as int;
        final type = PoseLandmarkType.fromLandmarkIndex(index);
        
        if (type != null) {
          landmarks[type] = PoseLandmark(
            type: type,
            // Flip X since we're using a mirrored frontend view but MediaPipe returns normal coords
            // Or keep as is, coordinate translator usually handles mirroring
            x: (lm['x'] as num).toDouble(),
            y: (lm['y'] as num).toDouble(),
            z: (lm['z'] as num).toDouble(),
            confidence: (lm['visibility'] as num).toDouble(),
          );
        }
      }
    }
    
    _frameCount++;
    _latestPose = PoseResult(
      landmarks: landmarks,
      frameNumber: _frameCount,
    );
    
    // Emit to stream
    _poseStreamController.add(_latestPose!);
  }

  @override
  Future<PoseResult?> detectPose(InputImageData imageData) async {
    // Not used - pose comes from Python backend stream
    return _latestPose;
  }

  @override
  Future<PoseResult?> processFrame(dynamic cameraImage) async {
    // Return the latest pose from the Python backend
    // The Python backend handles its own camera, so we don't process frames here
    return _latestPose;
  }
  
  /// Send ping to check connection
  void ping() {
    _sendCommand({'command': 'ping'});
  }

  @override
  void dispose() {
    _sendCommand({'command': 'stop'});
    _channel?.sink.close();
    _poseStreamController.close();
    _frameStreamController.close();
    _isReady = false;
  }
}

File: posture-coach-team-copy/frontend/services/mediapipe_pose_detector.ts
import { usePoseStore } from '../store/pose_state_manager';
import { PoseResult } from '../models/pose_result';
import { PoseLandmarkType, PoseLandmark } from '../models/pose_landmark';

export class MediaPipePoseDetector {
    private host: string;
    private port: number;
    private ws: WebSocket | null = null;
    private _isReady: boolean = false;
    private _frameCount: number = 0;

    constructor(host: string = 'localhost', port: number = 8765) {
        this.host = host;
        this.port = port;
    }

    get isReady(): boolean {
        return this._isReady;
    }

    async initialize(): Promise<void> {
        const { setConnectionStatus, setCurrentFrame, setError, handlePoseResult } = usePoseStore.getState();

        try {
            // In Android emulator, localhost is 10.0.2.2 usually.
            // Adjust if running on physical device or other environments.
            let finalHost = this.host;
            if (this.host === 'localhost') {
                // Platform.OS === 'android' ? '10.0.2.2' : 'localhost'
                // For Expo web/desktop it's localhost
                // We'll just assume localhost to match Flutter logic which ran on desktop
            }
            const uri = `ws://${finalHost}:${this.port}`;
            console.log(`🔌 Connecting to MediaPipe server at ${uri}`);

            this.ws = new WebSocket(uri);

            this.ws.onopen = () => {
                console.log('✅ Connected to MediaPipe server');
                this._isReady = true;
                setConnectionStatus(true, 'Connected to MediaPipe server');
                // No longer automatically starting Python webcam
            };

            this.ws.onmessage = (event) => {
                this.handleMessage(event.data);
            };

            this.ws.onerror = (error) => {
                console.log('❌ WebSocket error', error);
                this._isReady = false;
                setConnectionStatus(false, `Connection error`);
            };

            this.ws.onclose = () => {
                console.log('🔌 WebSocket connection closed');
                this._isReady = false;
                setConnectionStatus(false, 'Connection closed');
            };
        } catch (e: any) {
            console.log('❌ Failed to connect to MediaPipe server:', e);
            this._isReady = false;
            setConnectionStatus(false, `Failed to connect: ${e.message}`);
            throw e;
        }
    }

    private sendCommand(command: any) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify(command));
        }
    }

    private handleMessage(message: string) {
        try {
            const data = JSON.parse(message);
            const type = data.type;

            switch (type) {
                case 'frame':
                    if (data.pose) {
                        this.handlePoseData(data.pose);
                    }
                    if (data.image) {
                        usePoseStore.getState().setCurrentFrame(data.image);
                    }
                    break;
                case 'pose':
                    this.handlePoseData(data.data);
                    break;
                case 'status':
                    const success = data.success as boolean;
                    const msg = data.message as string;
                    console.log(`📊 Status: ${msg}`);
                    usePoseStore.getState().setConnectionStatus(success, msg);
                    break;
                case 'pong':
                    console.log('🏓 Pong received');
                    break;
            }
        } catch (e) {
            console.log('Error parsing message:', e);
        }
    }

    private handlePoseData(poseData: any) {
        const detected = poseData.detected ?? false;
        const landmarks = new Map<PoseLandmarkType, PoseLandmark>();

        if (detected && poseData.landmarks) {
            const landmarkList = poseData.landmarks;

            for (const lm of landmarkList) {
                const index = lm.index as number;
                // Map index to type since values are same
                const type = index as PoseLandmarkType;

                if (PoseLandmarkType[type] !== undefined) {
                    landmarks.set(type, new PoseLandmark(
                        type,
                        lm.x,
                        lm.y,
                        lm.z,
                        lm.visibility
                    ));
                }
            }
        }

        this._frameCount++;
        const latestPose = new PoseResult(landmarks, new Date(), this._frameCount);

        usePoseStore.getState().handlePoseResult(latestPose);
    }

    sendFrame(base64Image: string) {
        if (!this._isReady) return;
        this.sendCommand({ command: 'process_frame', image: base64Image, timestamp_ms: Date.now() });
    }

    sendMetrics(metricsPayload: any) {
        if (!this._isReady) return;
        this.sendCommand({ command: 'save_metrics', payload: metricsPayload });
    }

    startSession() {
        if (!this._isReady) return;
        this.sendCommand({ command: 'start' });
    }

    pauseSession() {
        if (!this._isReady) return;
        this.sendCommand({ command: 'pause' });
    }

    stopSession() {
        if (!this._isReady) return;
        this.sendCommand({ command: 'stop' });
    }

    ping() {
        this.sendCommand({ command: 'ping' });
    }

    dispose() {
        this.sendCommand({ command: 'stop' });
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
        this._isReady = false;
    }
}

export const poseDetectorService = new MediaPipePoseDetector('10.181.245.239', 8765);
